### ISSUE TITLE
Unable to uninstall Azure IoT Edge on Windows 10 Machine

### ISSUE THREAD

**Issue by rahulsehgal** [Azure/iotedge#2821]
Hi Team,

I am not able to uninstall Azure IoT Edge on Windows 10 Machine.(Machine is Azure VM)

I have tried all the options to delete all moby and data folders from ProgramData manually and run  Uninstall-IoTEdge command.

IoT Edge or the IoT Edge Moby Engine is installed in an invalid location. There may be an old preview install present. Please run Uninstall-IoTEdge first or reimage the device. To reinstall, first remove the existing installation using "Uninstall-IoTEdge".

As this is my DEV machine I also used edgehubdev on this machine as well. I tried deleting moby data folder but it is not allowing to delete the folder. I also tried renaming the folder still it has the same error. I am not able to install iot edge runtime on this machine. COuld you please suggest to install 1.0.9 version of iot edge runtime.

**Comment by massand** [comment]
Few questions: 
- Could you provide the full Uninstall-IoTEdge command you used, including parameters?
- Could you include the full output of that command?
- Are you running in elevated Powershell prompt mode?

**Comment by massand** [comment]
@rahulsehgal - I successfully ran the Uninstall-IoTEdge and Install-IoTEdge commands on a fresh VM. I encourage you to answer the questions above and step through the script's various legacy folders listed here to ensure everything is cleant up prior to running the script - https://github.com/Azure/iotedge/blob/master/scripts/windows/setup/IotEdgeSecurityDaemon.ps1#L51

If you still encounter issues, please let us know.

**Comment by rahulsehgal** [comment]
Hi Team,

Sorry for delay in reply. 
@massand 
I have used the following commands to unsintall 
Uninstall-IoTEdge
Output:
Uninstalling...
Deleting data directory 'C:\ProgramData\iotedge'...
Successfully uninstalled IoT Edge.

Call For Install 

. {Invoke-WebRequest -useb https://aka.ms/iotedge-win} | Invoke-Expression; `                      >> Deploy-IoTEdge                                              
Output :                                                         
ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     0.0        IoTEdge                             {Deploy-IoTEdge, Get-IoTEdgeLog, Import-IntermediateCertChai…

IoT Edge or the IoT Edge Moby Engine is installed in an invalid location. There may be an old preview install present. Please run Uninstall-IoTEdge first or reimage the device. To reinstall, first remove the existing installation using "Uninstall-IoTEdge".
ScriptHalted
At line:799 char:13
+             throw
+             ~~~~~
+ CategoryInfo          : OperationStopped: (:) [], RuntimeException
+ FullyQualifiedErrorId : ScriptHalted

Yes, I am using PS as an Administrator.

**Comment by rahulsehgal** [comment]
Another issue I see is in Program Data, there was a folder for iotedge-moby-data, which I am not able to delete as OS gives error while delete because it says files are in use.

I use this VM for DEV as well so I have intalled edgehubdev also.

To install runtime I have uninstalled docker, edgehubdev and also tried renaming  iotedge-moby-data to xxx. Still am unable to install iot edge.

**Comment by rahulsehgal** [comment]
@massand  please let me know if any other details are required.

I also tried installing the 1.0.7 though offline installer but it also gave me the same error.

{Invoke-WebRequest -useb aka.ms/iotedge-win} | Invoke-Expression; Deploy-IoTEdge -OfflineInstallationPath C:\offline\Microsoft-Azure-IoTEdge.cab  

Output :

StatusCode        : 200
StatusDescription : OK
Content           : {78, 101, 119, 45...}
RawContent        : HTTP/1.1 200 OK
                    x-amz-id-2: S/gcBqa1UiDpLfvzz6qEL0YWCFxbd6yvzLrgpD3q7sxXP8w2d9qe1XIsMa/5RY45Sm/HrhbQPEQ=
                    x-amz-request-id: 4252B057E53BDBD9
                    Content-Disposition: attachment; filename=IotEdgeSecurity...
Headers           : {[x-amz-id-2, S/gcBqa1UiDpLfvzz6qEL0YWCFxbd6yvzLrgpD3q7sxXP8w2d9qe1XIsMa/5RY45Sm/HrhbQPEQ=],
                    [x-amz-request-id, 4252B057E53BDBD9], [Content-Disposition, attachment;
                    filename=IotEdgeSecurityDaemon.ps1], [Accept-Ranges, bytes]...}
RawContentLength  : 91074


IoT Edge or the IoT Edge Moby Engine is installed in an invalid location. There may be an old preview install present. Please run Uninstall-IoTEdge first or reimage the device. To reinstall, first remove the existing installation using "Uninstall-IoTEdge".
ScriptHalted
At line:799 char:13
+             throw
+             ~~~~~
    + CategoryInfo          : OperationStopped: (:) [], RuntimeException
    + FullyQualifiedErrorId : ScriptHalted

**Comment by massand** [comment]
Could you run the follow Powershell commands (on a new elevated Powershell window) and provide the output here? Thanks!
```
Set-Variable LegacyEdgeInstallDirectory -Value 'C:\ProgramData\iotedge' -Option Constant
Set-Variable LegacyEdgeEventLogName -Value 'iotedged' -Option Constant
Set-Variable LegacyEdgeEventLogInstallDirectory -Value 'C:\ProgramData\iotedge-eventlog' -Option Constant
Set-Variable LegacyEventLogApplicationRegPath -Value 'HKLM:\SYSTEM\CurrentControlSet\Services\EventLog\Application' -Option Constant
Set-Variable LegacyMobyDataRootDirectory -Value "$env:ProgramData\iotedge-moby-data" -Option Constant
Set-Variable LegacyMobyStaticDataRootDirectory -Value 'C:\ProgramData\iotedge-moby-data' -Option Constant
Set-Variable LegacyMobyInstallDirectory -Value "$env:ProgramData\iotedge-moby" -Option Constant
Set-Variable LegacyMobyStaticInstallDirectory -Value 'C:\ProgramData\iotedge-moby' -Option Constant

Get-ChildItem $LegacyVarTest
Get-ChildItem $LegacyEdgeInstallDirectory
Get-ChildItem $LegacyEdgeEventLogName
Get-ChildItem $LegacyEdgeEventLogInstallDirectory
Get-ChildItem $LegacyEventLogApplicationRegPath
Get-ChildItem $LegacyMobyDataRootDirectory
Get-ChildItem $LegacyMobyStaticDataRootDirectory 
Get-ChildItem $LegacyMobyInstallDirectory 
Get-ChildItem $LegacyMobyStaticInstallDirectory

```

**Comment by massand** [comment]
Please let us know if you were able to resolve the issue or find the output of the command above.

**Comment by rahulsehgal** [comment]
@massand I executed the following commands 👍 
I got the result for the following two commands.

Get-ChildItem $LegacyVarTest
Directory: C:\WINDOWS\system32

Get-ChildItem $LegacyEventLogApplicationRegPath
 Hive: HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\EventLog\Application

![image](https://user-images.githubusercontent.com/18692450/81146220-26525980-8f78-11ea-95d8-1ebf8ba6238a.png)

**Comment by rahulsehgal** [comment]
Do we manually need to delete some of the Registry Entries.

**Comment by rahulsehgal** [comment]
@massand Is there any update on this issue. I am still facing this issue. Is there any way to Re-install iot-edge

**Comment by massand** [comment]
Your system looks clean as per my `Get-ChildItem` commands you ran. There are no registries to remove here. One more command to check - 
```
Get-Package 'microsoft-azure-iotedge'
```
If a package is returned, you will need to `Uninstall-Package 'microsoft-azure-iotedge'` first.

If that command doesn't return any package and this is a dev box, you could debug-step through the install PoSH script and determine which `Test-` function is failing around line numbers 947-978 in `scripts/windows/setup/IotEdgeSecurityDaemon.ps1`?

**Comment by ebertrams** [comment]
@rahulsehgal , have the recommendations from massand help you? Can we close this issue? Thanks

**Comment by ebertrams** [comment]
@rahulsehgal , any updates on this issue? Can we close this issue? Thanks!

**Comment by rahulsehgal** [comment]
@massand @ebertrams This issue is not resovled. I am not able to uninstall Old Runtime, tried the command mentioned aboved looked for PS script and no legacy directory exists.
'''
 if ($LegacyMobyStaticInstallDirectory -ne $LegacyMobyInstallDirectory) {

$legacyMobyData = (Test-Path -Path $LegacyMobyDataRootDirectory) -or
        (Test-Path -Path $LegacyMobyStaticDataRootDirectory)
'''

 And now we are facing similar issue in another VM.

**Comment by rahulsehgal** [comment]
Is there any way to clean-up existing Moby installation?

**Comment by massand** [comment]
> And now we are facing similar issue in another VM.

We need more information than that. Please attach / paste the complete legacy detection script you ran on both systems and attach/paste the output of the script. 

@xinshiMSFT - do you know if there are any collisions expected with installation of 1.0.9 on the same system as edgehubdev? It would be good to see if anyone can repro similar issues on the tooling side.

**Comment by micahl** [comment]
@rahulsehgal I'm closing this issue due to inactivity for now.  Reactivate if there's additional information.

**Comment by biaojiang** [comment]
1. Hi, I have the same problem of not being able to uninstall the IoT Edge in Windows 10 as well.

1.  First, I tried to uninstall it in Apps&Features. However, after that, the `Azure IoT Edge` app is still under "Apps", as shown in the following image:
![iot_edge_in_apps-and-features](https://user-images.githubusercontent.com/18609133/123213101-e3c64600-d4c5-11eb-9b31-0a92a655f36f.PNG)

2. Then I ran the command in PowerShell Admin, and got the following results:
![uninstall_iot_edge](https://user-images.githubusercontent.com/18609133/123213240-0e180380-d4c6-11eb-96f0-71c7eaf791a3.PNG)

3. Then I tried to create a new IoT Edge with `Windows Admin Center`, and got the following error:
![create_new_iot_edge_error](https://user-images.githubusercontent.com/18609133/123213713-9f877580-d4c6-11eb-915e-71bfecec8e1e.PNG)

Have no idea how to fully uninstall the Azure IoT Edge. Should I go to the `registry editor`? tks.
